Add VItemGroup component

sarayourfriend opened the pull request:
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->
Fixes #370 by @panchovm

## Description
<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->
Adds new `VItemGroup` and `VItem` components for rendering a grouped set of buttons.

This is currently blocked on focus styles for the items as the current focus styles won't work:

<img width="328" alt="Captura de Tela 2021-11-11 às 12 59 57" src="https://user-images.githubusercontent.com/24264157/141368279-3fff0693-84c3-4e4c-8126-373dccb04ba4.png">

It's also currently missing the "direction" behavior to render a horizontal list, useful for https://github.com/WordPress/openverse-frontend/issues/329

## Testing Instructions
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete this section entirely. -->
Run `npm run storybook` and visit `http://localhost:6006/?path=/story/components-vitemgroup--default`.

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. -->
- [ ] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [ ] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

sarayourfriend commented:
@panchovm whenever you get a chance, the current focus styles are show in the screenshot in the PR description. Would love some advice on how to fix this. One option I had was to raise the z-index of the focused item so that the focus styles will show over adjacent items. I think this looks quite nice but I wanted to verify it with you before committing to it:

<img width="321" alt="Captura de Tela 2021-11-11 às 13 16 33" src="https://user-images.githubusercontent.com/24264157/141370215-7df78d1f-9b7d-452a-9154-ab49c16a0b22.png">

sarayourfriend commented:
This is fully ready for review aside from the focus styles issue which should hopefully be simple to implement anyway.

sarayourfriend commented:
Adding a video to help show @panchovm the current behavior due to local environment issues:

https://user-images.githubusercontent.com/24264157/141479723-741748aa-8ce6-4440-9e84-b0bf79a87704.mov

I'm noticing a couple issues:

1. Duplicated top border on items other than the first one
2. Each item "jiggles" when it is focused

panchovm commented:
Could this be related to #370 and [this answer](https://github.com/WordPress/openverse-frontend/issues/370#issuecomment-967140504)? I'm seeing the mockups and here are the two uses I found.

| Audio format | Mobile menu |
| ----------- | ----------- |
| <img width="300" alt="Screen Shot 2021-11-12 at 11 04 50" src="https://user-images.githubusercontent.com/895819/141479738-e1bd7e46-2934-4c7e-8442-a71a4a7b7439.png"> | <img width="300" alt="Screen Shot 2021-11-12 at 11 05 47" src="https://user-images.githubusercontent.com/895819/141479744-2bf3c323-72d5-49b8-b18b-25343a1237e9.png"> |

sarayourfriend commented:
I think we'll need to create a slightly separate implementation for the popover contents rather than standalone grouped items like this.

Here're the updated focus styles:

https://user-images.githubusercontent.com/24264157/141483847-b2ffc741-3a95-42db-8e09-7736f5fa6d6c.mov

sarayourfriend commented:
Okay, I fixed some issues and added a borderless design that will allow this component to be used inside and outside of a popover 🙂 

https://user-images.githubusercontent.com/24264157/141490752-e1b40392-387d-404e-bd6f-efbb63a28a5a.mov

panchovm commented:
It looks wonderful. Great job 👏

obulat commented on src/components/VIcon/VIcon.vue:
```diff
@@ -29,7 +29,10 @@ export default {
+      /**
+       * In `just` our icons get transformed to Vue components
+       */
+      type: process.env.NODE_ENV === 'test' ? Object : String,
```
Thank you for explaining this!

zackkrida replied:
This comment should read `jest`, not `just`, right?

sarayourfriend replied:
Yup!

obulat commented on src/components/VItemGroup/VItemGroup.vue:
```diff
@@ -0,0 +1,76 @@
+     * @default 'vertical'
+     */
+    direction: {
+      type: /** @type {import('@nuxtjs/composition-api').PropType<'vertical' | 'horizontal'>} */ (String),
```
I didn't know you can type the props this way :)

obulat commented on src/assets/icons/check.svg:
```diff
@@ -0,0 +1,3 @@
+<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none">
```
I think the same icon is used in the VCheckbox component.

obulat reviewed (approved):
I love the way the components work and look, and they sound great in Voice Over. There are so many code patterns to learn from :)

zackkrida commented:
This is so awesome, some thoughts/questions.

-  I think `z-index: 1;` on the focused element would be ideal; I see a small 1/1.5px imperfection on the focus ring without it

    > before

    ![CleanShot 2021-11-19 at 08 01 34@2x](https://user-images.githubusercontent.com/6351754/142627124-3a4fe2b9-b886-4292-952f-6a58fe7208c0.png)

    > after

    ![CleanShot 2021-11-19 at 08 01 54@2x](https://user-images.githubusercontent.com/6351754/142627153-aa70aff3-8a61-491b-ad69-8c6044ecfba9.png)

- Some `radiogroup` [implementations](https://www.w3.org/TR/2016/WD-wai-aria-practices-1.1-20160317/examples/radio/radio.html) I've tested use `tab` to focus in and out of the entire group, but use keydown/up/left/right to cycle through the radio options. Have you considered implementing that here? Alternatively, I think it could just be an aria `group` instead of a radiogroup?
- There's a visual bug in the horizontal popover example, where each item is missing the correct 'end' padding:
    ![CleanShot 2021-11-19 at 08 12 09@2x](https://user-images.githubusercontent.com/6351754/142628642-dc764623-7490-440c-9669-8283bfdf470c.png)

sarayourfriend commented:
> Some radiogroup implementations I've tested use tab to focus in and out of the entire group, but use keydown/up/left/right to cycle through the radio options. Have you considered implementing that here? Alternatively, I think it could just be an aria group instead of a radiogroup?

That's a good point, I completely forgot to implement that interaction! Our DropdownMenu component already does that anyway 🙂

sarayourfriend commented:
I don't think the group role works for this case, BTW. I'll work on fixing the radiogroup implementation 👍 

https://www.w3.org/TR/wai-aria-1.1/#group

sarayourfriend commented on src/components/VItemGroup/VItemGroup.vue:
```diff
@@ -70,7 +94,69 @@ export default defineComponent({
+
+      const targetIndex = items.findIndex((item) => item === target)
+
+      switch (event.key) {
+        case 'ArrowUp':
+        case 'ArrowLeft':
+          if (targetIndex === 0) {
+            // If navigating up/left on the first item, ignore it
+            return
+          }
+          ensureFocus(items[targetIndex - 1])
+          break
+        case 'ArrowDown':
+        case 'ArrowRight':
+          if (targetIndex === items.length - 1) {
+            // If navigation down/right on the last item, ignore it
+            return
+          }
+          ensureFocus(items[targetIndex + 1])
+          break
+      }
```
I think this might need some tweaking for RTL support with the Left/Right handlers.

zackkrida replied:
s/could down/right on the last item actually cycle back to the first item?

sarayourfriend replied:
Ah you're right, it should. I missed that in the example.

sarayourfriend replied:
Fixed!
